promql/slow_query_logger.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package promql
+
+import (
+	"io"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/config"
+)
+
+func (ng *Engine) updateSlowQueryLogger(c *config.Config) error {
+	if d := c.GlobalConfig.QueryLogMinDuration; d != nil {
+		du := time.Duration(*d)
+		ng.queryLogMinDuration = &du
+		oldFile := ng.slowQueriesFile
+		// We always open the file again; so that a Prometheus reload can be
+		// called by external programs when they rotate the slow log file.
+		f, err := os.OpenFile(c.GlobalConfig.QueryLogFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			return errors.Wrap(err, "can't create slow query log")
+		}
+		ng.slowQueriesFile = f
+		ng.slowQueriesLogger = newSlowQueriesLogger(f)
+		if oldFile != nil {
+			oldFile.Close()
+		}
+	} else {
+		ng.queryLogMinDuration = nil
+		ng.slowQueriesLogger = nil
+		if ng.slowQueriesFile != nil {
+			ng.slowQueriesFile.Close()
+			ng.slowQueriesFile = nil
+		}
+	}
+	return nil
+}
+
+func newSlowQueriesLogger(w io.Writer) log.Logger {
+	var (
+		timestampFormat = log.TimestampFormat(
+			func() time.Time { return time.Now().UTC() },
+			"2006-01-02T15:04:05.000Z07:00",
+		)
+	)
+	l := log.NewLogfmtLogger(log.NewSyncWriter(w))
```

Comment:
This is a hot path, I'm not sure no buffering is wise. We should probably also have this as JSON only.

Reply:
done